dali/pipeline/operators/arithmetic/expression_tree.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_cpu.h"
+
+namespace dali {
+
+namespace {
+
+// <expr> := <call>|<scalar>|<input>
+// <call> := <name> '(' <expr>* ')'
+// <name> := fun
+// <input> := &Int
+// <scalar> := $Int:Type
+
+// add(&0 mul(&1 $0:i8))
+// add(&0 rand())
+
+std::tuple<std::unique_ptr<ExpressionBase>, int> ParseExpr(const std::string &expr, int pos);
```

Comment:
Why restrict this to std::string (and not `const char *`)? Alternatively, you could use a stream "concept" -  something that provides the functions:
```
get()
peek()
putback()
```
You could use it with `std::istringstream`.

Reply:
Because it was pretty straightforward to write, although I first looked at string_view (not until c++17), and because this is the type we get this description in.